src/Microsoft.ML.Transforms/TensorflowTransform.cs

```diff
@@ -0,0 +1,681 @@
+            var input = Source.GetRowCursor(inputPred, rand);
+            return new RowCursor(Host, this, input, active);
+        }
+
+        public override IRowCursor[] GetRowCursorSet(out IRowCursorConsolidator consolidator, Func<int, bool> predicate, int n, IRandom rand = null)
+        {
+            Host.CheckValue(predicate, nameof(predicate));
+            Host.CheckValueOrNull(rand);
+
+            var inputPred = _bindings.GetDependencies(predicate);
+            var active = _bindings.GetActive(predicate);
+            var inputs = Source.GetRowCursorSet(out consolidator, inputPred, n, rand);
+            Host.AssertNonEmpty(inputs);
+
+            if (inputs.Length == 1 && n > 1 && _bindings.AnyNewColumnsActive(predicate))
+                inputs = DataViewUtils.CreateSplitCursors(out consolidator, Host, inputs[0], n);
+            Host.AssertNonEmpty(inputs);
+
+            var cursors = new IRowCursor[inputs.Length];
+            for (int i = 0; i < inputs.Length; i++)
+                cursors[i] = new RowCursor(Host, this, inputs[i], active);
+            return cursors;
+        }
+
+        private interface ITensorValueGetter
+        {
+            TFTensor GetTensor();
+        }
+
+        private class TensorValueGetter<T> : ITensorValueGetter
+        {
+            private readonly ValueGetter<T> _srcgetter;
+            private readonly ColumnType _type;
+            private readonly ConstructorInfo _ctor;
+
+            public TensorValueGetter(IRow input, int colIndex)
+            {
+                _srcgetter = input.GetGetter<T>(colIndex);
+                _type = input.Schema.GetColumnType(colIndex);
+                _ctor = typeof(TFTensor).GetConstructor(new[] { typeof(T) });
+            }
+            public TFTensor GetTensor()
+            {
+                var scalar = default(T);
+                _srcgetter(ref scalar);
+                return (TFTensor)_ctor.Invoke(null, new object[] { scalar });
+            }
+        }
+
+        private class TensorValueGetterVec<T> : ITensorValueGetter
+        {
+            private readonly ValueGetter<VBuffer<T>> _srcgetter;
+            private readonly ColumnType _type;
+            private readonly TFShape _tfShape;
+            private readonly MethodInfo _method;
+            public TensorValueGetterVec(IRow input, int colIndex, TFShape tfShape)
+            {
+                _srcgetter = input.GetGetter<VBuffer<T>>(colIndex);
+                _type = input.Schema.GetColumnType(colIndex);
+                _tfShape = tfShape;
+                _method = typeof(TFTensor).GetMethod("FromBuffer",
+                        BindingFlags.Public | BindingFlags.Static,
+                        Type.DefaultBinder,
+                        new[] { typeof(TFShape), typeof(T[]), typeof(int), typeof(int) },
+                        null);
+            }
+            public TFTensor GetTensor()
+            {
+                var tmpBuf = default(VBuffer<T>);
```

Comment:
>tmpBuf [](start = 20, length = 6)

This class should have a VBuffer as a private field, otherwise a new array will be allocated on every call.

Reply:
Alternatively, you can make ITensorValueGetter generic on "T", and the GetTensor method will be GetTensor(ref T input).

---
In reply to: [210415541](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210415541) [](ancestors = 210415541)